File: Review-07-Java-Toys/Example-3/src/test/java/edu/odu/cs/cs350/examples/TestRoster.java
package edu.odu.cs.cs350.examples;

import org.junit.FixMethodOrder;
import org.junit.runners.MethodSorters;
import org.junit.Test;
import org.junit.Before;

import static org.junit.Assert.*;

import static org.hamcrest.CoreMatchers.*;
import org.hamcrest.core.IsNull;
import org.hamcrest.core.IsSame;

/**
 * 1 - Does this piece of code perform the operations 
 *     it was designed to perform?
 * 
 * 2 - Does this piece of code do something it was not 
 *     designed to perform?
 * 
 * 1 Test per mutator
 */
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestRoster {
    // Students - will not be changed during tests
    Student john  = new Student("John");
    Student tom   = new Student("Tom");
    Student jay   = new Student("Jay");
    Student oscar = new Student("Oscar");

    // Student "lists" - will not be changed during tests
    Student[] firstThreeStudents = {john, tom, jay};
    Student[] allStudents        = {john, tom, jay, oscar};

    // Course Rosters - will possibly change in each test
    Roster defaultCourse;
    Roster emptyCS350;

    @Before
    public void setUp()
    {
        defaultCourse = new Roster();
        emptyCS350    = new Roster(3, "CS 350");
    }

    @Test
    public void testDefaultConstructor()
    {      
        assertThat(defaultCourse.getCourseNum(), equalTo("CS 150"));
        assertThat(defaultCourse.getEnrollLimit(),
                   equalTo(Roster.DEFAULT_MAX_STUDENTS));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(defaultCourse.listEnrolledStudents(), notNullValue());
        assertThat(defaultCourse.listEnrolledStudents().size(), equalTo(0));

        // skipping hashcode
        // skipping equals

        // test toString
        assertThat(defaultCourse.toString(), containsString("CS 150"));
        assertThat(defaultCourse.toString(),
                   containsString(Integer.toString(defaultCourse.numEnrolled())));
        assertThat(defaultCourse.toString(),
                   containsString(Integer.toString(Roster.DEFAULT_MAX_STUDENTS)));
    }

    @Test
    public void testNonDefaultConstructor()
    {
        assertThat(emptyCS350.getCourseNum(), equalTo("CS 350"));
        assertThat(emptyCS350.getEnrollLimit(), equalTo(3));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(emptyCS350.listEnrolledStudents(), notNullValue());
        assertThat(emptyCS350.listEnrolledStudents().size(), equalTo(0));

        // NOT skipping hashcode
        assertThat(emptyCS350.hashCode(), not(equalTo(defaultCourse.hashCode())));
        // NOT skipping equals
        assertThat(emptyCS350, not(equalTo(defaultCourse)));

        // test toString
        assertThat(emptyCS350.toString(), containsString("CS 350"));
        assertThat(emptyCS350.toString(),
                   containsString(Integer.toString(emptyCS350.numEnrolled())));
        assertThat(emptyCS350.toString(), containsString(Integer.toString(3)));
    }

    @Test
    public void testSetCourseNum()
    {
        Roster cs252 = new Roster();

        int oldHashCode = cs252.hashCode();

        cs252.setCourseNum("CS 252");

        assertThat(cs252.getCourseNum(), equalTo("CS 252"));
        assertThat(cs252.getEnrollLimit(),
                   equalTo(Roster.DEFAULT_MAX_STUDENTS));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(cs252.listEnrolledStudents(), notNullValue());
        assertThat(cs252.listEnrolledStudents().size(), equalTo(0));

        // NOT skipping hashcode
        assertThat(cs252.hashCode(), not(equalTo(oldHashCode)));
        // NOT skipping equals
        assertThat(cs252, not(equalTo(defaultCourse)));

        // test toString
        assertThat(cs252.toString(), containsString("CS 252"));
        assertThat(cs252.toString(),
                   containsString(Integer.toString(cs252.numEnrolled()))); // fixed mistake
        assertThat(cs252.toString(), 
                   containsString(Integer.toString(Roster.DEFAULT_MAX_STUDENTS)));
    }

    @Test
    public void testSetEnrollLimit()
    {       
        emptyCS350.setEnrollLimit(2);

        assertThat(emptyCS350.getCourseNum(), equalTo("CS 350"));
        assertThat(emptyCS350.getEnrollLimit(), equalTo(2));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(emptyCS350.listEnrolledStudents(), notNullValue());
        assertThat(emptyCS350.listEnrolledStudents().size(), equalTo(0));

        // NOT skipping hashcode
        assertThat(emptyCS350.hashCode(), not(equalTo(defaultCourse.hashCode())));
        // NOT skipping equals
        assertThat(emptyCS350, not(equalTo(defaultCourse)));

        // test toString
        assertThat(emptyCS350.toString(), containsString("CS 350"));
        assertThat(emptyCS350.toString(),
                   containsString(Integer.toString(emptyCS350.numEnrolled())));
        assertThat(emptyCS350.toString(), containsString(Integer.toString(2)));
    }

    @Test
    public void testEnroll()
    {
        //This is where the fun starts
        Roster cs725 = new Roster(3, "CS 725");

        int oldHashCode = cs725.hashCode();

        // try to add 4 students
        assertThat(cs725.enroll(john), is(true));
        assertThat(cs725.enroll(tom), is(true));
        assertThat(cs725.enroll(jay), is(true));
        assertThat(cs725.enroll(oscar), is(false)); // should fail (limit of 3)

        assertThat(cs725.getCourseNum(), equalTo("CS 725"));
        assertThat(cs725.getEnrollLimit(), equalTo(3));
        assertThat(cs725.numEnrolled(), equalTo(3)); // fixed mistake

        // 3 students have been added
        assertThat(cs725.listEnrolledStudents(), notNullValue());
        assertThat(cs725.listEnrolledStudents().size(), equalTo(3));
        assertThat(cs725.listEnrolledStudents().toArray(),
                   equalTo(firstThreeStudents));

        assertThat(cs725.hashCode(), not(equalTo(oldHashCode)));

        oldHashCode = cs725.hashCode();

        // Change the limit to 4
        cs725.setEnrollLimit(4); // mistake - this should be cs725
        assertThat(cs725.getEnrollLimit(), equalTo(4)); // mistake - this should be cs725

        // try to add a 4th student with the new limit of 4
        assertThat(cs725.enroll(oscar), is(true)); // should succeed (limit of 4)

        // 4 students have been added
        assertThat(cs725.listEnrolledStudents(), notNullValue());
        assertThat(cs725.listEnrolledStudents().size(), equalTo(4));
        assertThat(cs725.listEnrolledStudents().toArray(),
                   equalTo(allStudents));

        // NOT skipping hashcode
        assertThat(cs725.hashCode(), not(equalTo(oldHashCode)));
        // NOT skipping equals
        assertThat(cs725, not(equalTo(defaultCourse)));

        // **test flaw** - did not exercise adding the same student twice
        assertThat(cs725.enroll(tom), is(false)); // should fail

        // test toString
        assertThat(cs725.toString(), containsString("CS 725"));
        assertThat(cs725.toString(),
                   containsString(Integer.toString(cs725.numEnrolled())));
        assertThat(cs725.toString(), 
                   containsString(Integer.toString(4)));

        assertThat(cs725.toString(), containsString(allStudents[0].toString()));
        assertThat(cs725.toString(), containsString(allStudents[1].toString()));
        assertThat(cs725.toString(), containsString(allStudents[2].toString()));
        assertThat(cs725.toString(), containsString(allStudents[3].toString()));
    }

    @Test
    public void testClone()
    {
        //This is where the fun continues
        Roster cs725 = new Roster(3, "CS 725");

        cs725.enroll(john);
        cs725.enroll(tom);
        cs725.enroll(jay);
      
        // Make the copy
        Roster copyCs725 = (Roster) cs725.clone();

        // Both Rosters should still be identical
        assertThat(copyCs725.getCourseNum(), equalTo(cs725.getCourseNum()));
        assertThat(copyCs725.getEnrollLimit(), equalTo(cs725.getEnrollLimit()));
        assertThat(copyCs725.numEnrolled(), equalTo(cs725.numEnrolled()));
        assertThat(copyCs725.hashCode(), equalTo(cs725.hashCode()));
        assertThat(copyCs725, equalTo(cs725));
        assertThat(copyCs725.toString(), equalTo(cs725.toString()));

        // But distinct
        assertThat(copyCs725, not(sameInstance(cs725)));
        assertThat(copyCs725.listEnrolledStudents(),
                   not(sameInstance(cs725.listEnrolledStudents())));

        // Change the limit to 4
        copyCs725.setEnrollLimit(4);
        assertThat(copyCs725.getEnrollLimit(), equalTo(4));
        assertThat(copyCs725.enroll(oscar), is(true));

        assertThat(copyCs725.listEnrolledStudents(), notNullValue());
        assertThat(copyCs725.listEnrolledStudents().size(), equalTo(4));
        assertThat(copyCs725.listEnrolledStudents().toArray(), equalTo(allStudents));

        assertThat(copyCs725.hashCode(), not(equalTo(cs725.hashCode())));
        assertThat(copyCs725, not(equalTo(defaultCourse)));
        assertThat(copyCs725, not(equalTo(cs725)));

        // cs725 should be unchanged
        assertThat(cs725.listEnrolledStudents().size(), equalTo(3));
        assertThat(cs725.listEnrolledStudents().toArray(),
                   equalTo(firstThreeStudents));

        // test toString
        assertThat(copyCs725.toString(), containsString("CS 725"));
        assertThat(copyCs725.toString(),
                   containsString(Integer.toString(copyCs725.numEnrolled())));
        assertThat(copyCs725.toString(), 
                   containsString(Integer.toString(4)));

        assertThat(copyCs725.toString(), containsString(allStudents[0].toString()));
        assertThat(copyCs725.toString(), containsString(allStudents[1].toString()));
        assertThat(copyCs725.toString(), containsString(allStudents[2].toString()));
        assertThat(copyCs725.toString(), containsString(allStudents[3].toString()));

        assertThat(copyCs725.toString(), not(equalTo(cs725.toString())));
    }
}

File: Review-07-Comparing-Languages/Example-3-Python/tests/examples/test_roster.py

import copy
import unittest
from hamcrest import *

from examples.student import Student
from examples.roster import (Roster, DEFAULT_MAX_STUDENTS)


class TestRoster(unittest.TestCase):
    """
    1 - Does this piece of code perform the operations
        it was designed to perform?

    2 - Does this piece of code do something it was not
        designed to perform?

    1 Test per mutator
    """

    # Students - will not be changed during tests
    john = Student("John")
    tom = Student("Tom")
    jay = Student("Jay")
    oscar = Student("Oscar")

    # Student "lists" - will not be changed during tests
    all_students = [john, tom, jay, oscar]
    first_three_students = all_students[:3]


    def setUp(self):

        self.default_course = Roster()
        self.empty_cs350 = Roster(3, "CS 350")

    def test_default_constructor(self):

        assert_that(self.default_course.course_num, equal_to("CS 150"))
        assert_that(self.default_course.enroll_limit,
                    equal_to(DEFAULT_MAX_STUDENTS))

        assert_that(self.default_course.num_enrolled(), equal_to(0))

        # No students have been added
        assert_that(self.default_course.list_enrolled_students(), not_none())
        assert_that(len(self.default_course.list_enrolled_students()), equal_to(0))

        # skipping hashcode
        # skipping equals

        # test to_string
        assert_that(str(self.default_course), contains_string("CS 150"))
        assert_that(str(self.default_course),
                    contains_string(f"{self.default_course.num_enrolled()}"))
        assert_that(str(self.default_course),
                    contains_string(f"{DEFAULT_MAX_STUDENTS}"))


    def test_non_default_constructor(self):
        assert_that(self.empty_cs350.course_num, equal_to("CS 350"))
        assert_that(self.empty_cs350.enroll_limit, equal_to(3))

        assert_that(self.default_course.num_enrolled(), equal_to(0))

        # No students have been added
        assert_that(self.empty_cs350.list_enrolled_students(), not_none())
        assert_that(len(self.empty_cs350.list_enrolled_students()), equal_to(0))

        # NOT skipping hashcode
        assert_that(hash(self.empty_cs350), is_not(equal_to(hash(self.default_course))))
        # NOT skipping equals
        assert_that(self.empty_cs350, is_not(equal_to(self.default_course)))

        # test to_string
        assert_that(str(self.empty_cs350), contains_string("CS 350"))
        assert_that(str(self.empty_cs350),
                    contains_string(f"{self.empty_cs350.num_enrolled()}"))
        assert_that(str(self.empty_cs350), contains_string("3"))

    def test_set_course_num(self):

        cs252 = Roster()

        old_hash_code = hash(cs252)

        cs252.course_num = "CS 252"

        assert_that(cs252.course_num, equal_to("CS 252"))
        assert_that(cs252.enroll_limit,
                    equal_to(DEFAULT_MAX_STUDENTS))

        assert_that(self.default_course.num_enrolled(), equal_to(0))

        # No students have been added
        assert_that(cs252.list_enrolled_students(), not_none())
        assert_that(len(cs252.list_enrolled_students()), equal_to(0))

        # NOT skipping hashcode
        assert_that(hash(cs252), is_not(equal_to(old_hash_code)))
        # NOT skipping equals
        assert_that(cs252, is_not(equal_to(self.default_course)))

        # test to_string
        assert_that(str(cs252), contains_string("CS 252"))
        assert_that(str(cs252),
                    contains_string(f"{cs252.num_enrolled()}")) # fixed mistake
        assert_that(str(cs252), contains_string(f"{DEFAULT_MAX_STUDENTS}"))


    def test_set_enroll_limit(self):

        self.empty_cs350.enroll_limit = 2

        assert_that(self.empty_cs350.course_num, equal_to("CS 350"))
        assert_that(self.empty_cs350.enroll_limit, equal_to(2))

        assert_that(self.default_course.num_enrolled(), equal_to(0))

        # No students have been added
        assert_that(self.empty_cs350.list_enrolled_students(), not_none())
        assert_that(len(self.empty_cs350.list_enrolled_students()), equal_to(0))

        # NOT skipping hashcode
        assert_that(hash(self.empty_cs350), is_not(equal_to(hash(self.default_course))))
        # NOT skipping equals
        assert_that(self.empty_cs350, is_not(equal_to(self.default_course)))

        # test to_string
        assert_that(str(self.empty_cs350), contains_string("CS 350"))
        assert_that(str(self.empty_cs350),
                    contains_string(f"{self.empty_cs350.num_enrolled()}"))
        assert_that(str(self.empty_cs350), contains_string("2"))

    def test_enroll(self):

        #This is where the fun starts
        cs725 = Roster(3, "CS 725")

        old_hash_code = hash(cs725)

        # try to add 4 students
        assert_that(cs725.enroll(TestRoster.john), is_(True))
        assert_that(cs725.enroll(TestRoster.tom), is_(True))
        assert_that(cs725.enroll(TestRoster.jay), is_(True))
        assert_that(cs725.enroll(TestRoster.oscar), is_(False)) # should fail (limit of 3)

        assert_that(cs725.course_num, equal_to("CS 725"))
        assert_that(cs725.enroll_limit, equal_to(3))
        assert_that(cs725.num_enrolled(), equal_to(3)) # fixed mistake

        # 3 students have been added
        assert_that(cs725.list_enrolled_students(), not_none())
        assert_that(len(cs725.list_enrolled_students()), equal_to(3))
        assert_that(cs725.list_enrolled_students(),
                    contains_inanyorder(*TestRoster.first_three_students))

        assert_that(hash(cs725), is_not(equal_to(old_hash_code)))

        old_hash_code = hash(cs725)

        # Change the limit to 4
        cs725.set_enroll_limit(4) # mistake - this should be cs725
        assert_that(cs725.enroll_limit, equal_to(4)) # mistake - this should be cs725

        # try to add a 4th student with the new limit of 4
        assert_that(cs725.enroll(TestRoster.oscar), is_(True)) # should succeed (limit of 4)

        # 4 students have been added
        assert_that(cs725.list_enrolled_students(), not_none())
        assert_that(len(cs725.list_enrolled_students()), equal_to(4))
        assert_that(cs725.list_enrolled_students(),
                    contains_inanyorder(*TestRoster.all_students))

        # NOT skipping hashcode
        assert_that(hash(cs725), is_not(equal_to(old_hash_code)))
        # NOT skipping equals
        assert_that(cs725, is_not(equal_to(self.default_course)))

        # **test flaw** - did not exercise adding the same student twice
        assert_that(cs725.enroll(TestRoster.tom), is_(False)) # should fail

        # test to_string
        assert_that(str(cs725), contains_string("CS 725"))
        assert_that(str(cs725), contains_string(f"{cs725.num_enrolled()}"))
        assert_that(str(cs725), contains_string("4"))

        assert_that(str(cs725), contains_string(str(TestRoster.all_students[0])))
        assert_that(str(cs725), contains_string(str(TestRoster.all_students[1])))
        assert_that(str(cs725), contains_string(str(TestRoster.all_students[2])))
        assert_that(str(cs725), contains_string(str(TestRoster.all_students[3])))

    def test_clone(self):

        #This is where the fun continues
        cs725 = Roster(3, "CS 725")

        cs725.enroll(TestRoster.john)
        cs725.enroll(TestRoster.tom)
        cs725.enroll(TestRoster.jay)

        # Make the copy
        copy_cs725 = copy.deepcopy(cs725)

        # Both Rosters should still be identical
        assert_that(copy_cs725.course_num, equal_to(cs725.course_num))
        assert_that(copy_cs725.enroll_limit, equal_to(cs725.enroll_limit))
        assert_that(copy_cs725.num_enrolled(), equal_to(cs725.num_enrolled()))
        assert_that(hash(copy_cs725), equal_to(hash(cs725)))
        assert_that(copy_cs725, equal_to(cs725))
        assert_that(str(copy_cs725), equal_to(str(cs725)))

        # But distinct
        assert_that(copy_cs725, is_not(same_instance(cs725)))
        assert_that(copy_cs725.list_enrolled_students(),
                    is_not(same_instance(cs725.list_enrolled_students())))

        # Change the limit to 4
        copy_cs725.set_enroll_limit(4)
        assert_that(copy_cs725.enroll_limit, equal_to(4))
        assert_that(copy_cs725.enroll(TestRoster.oscar), is_(True))

        assert_that(copy_cs725.list_enrolled_students(), not_none())
        assert_that(len(copy_cs725.list_enrolled_students()), equal_to(4))
        assert_that(copy_cs725.list_enrolled_students(),
                    contains_inanyorder(*TestRoster.all_students))

        assert_that(hash(copy_cs725), is_not(equal_to(hash(cs725))))
        assert_that(copy_cs725, is_not(equal_to(self.default_course)))
        assert_that(copy_cs725, is_not(equal_to(cs725)))

        # cs725 should be unchanged
        assert_that(len(cs725.list_enrolled_students()), equal_to(3))
        assert_that(cs725.list_enrolled_students(),
                    contains_inanyorder(*TestRoster.first_three_students))

        # test to_string
        assert_that(str(copy_cs725), contains_string("CS 725"))
        assert_that(str(copy_cs725), contains_string(f"{copy_cs725.num_enrolled()}"))
        assert_that(str(copy_cs725), contains_string("4"))

        assert_that(str(copy_cs725), contains_string(str(TestRoster.all_students[0])))
        assert_that(str(copy_cs725), contains_string(str(TestRoster.all_students[1])))
        assert_that(str(copy_cs725), contains_string(str(TestRoster.all_students[2])))
        assert_that(str(copy_cs725), contains_string(str(TestRoster.all_students[3])))

        assert_that(str(copy_cs725), is_not(str(equal_to(cs725))))
